uqcsbot/morse.py

```diff
@@ -0,0 +1,159 @@
+
+from uqcsbot.bot import UQCSBot
+
+# value of all valid ascii values in morse code
+MorseCodeDict = {
+    "A": ". - ",
+    "B": "- . . . ",
+    "C": "- . - . ",
+    "D": "- . . ",
+    "E": ". ",
+    "F": ". . - . ",
+    "G": "- - . ",
+    "H": ". . . . ",
+    "I": ". . ",
+    "J": ". - - - ",
+    "K": "- . - ",
+    "L": ". - . . ",
+    "M": "- - ",
+    "N": "- . ",
+    "O": "- - - ",
+    "P": ". - - . ",
+    "Q": "- - . - ",
+    "R": ". - . ",
+    "S": ". . . ",
+    "T": "- ",
+    "U": ". . - ",
+    "V": ". . . - ",
+    "W": ". - - ",
+    "X": "- . . - ",
+    "Y": "- . - - ",
+    "Z": "- - . . ",
+    "1": ". - - - - ",
+    "2": ". . - - - ",
+    "3": ". . . - - ",
+    "4": ". . . . - ",
+    "5": ". . . . . ",
+    "6": "- . . . . ",
+    "7": "- - . . . ",
+    "8": "- - - . . ",
+    "9": "- - - - . ",
+    "0": "- - - - - ",
+    ",": "- - . . - - ",
+    ".": ". - . - . - ",
+    "?": ". . - - . . ",
+    "/": "- . . - . ",
+    "-": "- . . . . - ",
+    "(": "- . - - . ",
+    ")": "- . - - . - ",
+    " ": "  ",
+    ";": "_ . _ . _ . ",
+    ":": "_ _ _ . . .",
+    "'": ". _ _ _ _ . ",
+    '"': ". _ . . _ . ",
+    "_": ". . _ _ . _ ",
+    "=": "_ . . . _ ",
+    "+": ". _ . _ . ",
+}
+
+
+class morse(commands.Cog):
+    def __init__(self, bot: UQCSBot):
+        self.bot = bot
+
+    @app_commands.command(name="morse")
+    @app_commands.describe(message="The message to be converted to morse code")
+    async def morse_command(
+        self, interaction: discord.Interaction, message: str
+    ) -> None:
+        """
+        Returns a string containing the argument converted to dots and dashes.
+        """
+
+        # Sanitise invalid characters from the message
+        message = morse.sanitise_illegals(message)
+
+        # Sanitise message for discord emotes
+        message = morse.sanitise_emotes(message)
+
+        # Convert all chars to lower case
+        message = message.upper()
+
+        # Then check they are valid morse code ascii
+        invalid = morse.check(message)
+        if invalid != 0:
+            await interaction.response.send_message(
+                f"Invalid morse code character/s in string: {invalid}"
+            )
+            return
+
+        # encrypt message
+        response = morse.encrypt_to_morse(message)
+
+        # Check if response is too long
+        if len(response) > 2000:
+            await interaction.response.send_message("Resulting message too long!")
+            return
+
+        await interaction.response.send_message(response)
+
+    @staticmethod
+    def sanitise_illegals(message: str) -> str:
+        """
+        Replaces all illegal characters in the message with their sanitised
+        equivalent.
+        """
+
+        # Strip whitespace from either side of the message
+        message = message.strip()
+
+        # replace code blocks with their sanitised equivalent
+        message = message.replace("```", "'''")
```

Comment:
Haven't tested - is it worth just stripping all ` characters? If code blocks have issues, inline code will probably still have issues.

Reply:
Just copied it from cowsay. Didn't actually read it. If a user prints something that's code, they probably just want the morse of the actual code, so I've replaced "`" with "".

Reply:
It wouldn't be unwise to move "sanitisation" like this out to a set of functions that can be used bot-wide, instead of re-implementing them every cog, but that's out-of-scope here. Sounds good to me :+1: